mruby/src/value/mod.rs

```diff
@@ -83,20 +81,18 @@ impl Value {
 impl Clone for Value {
     fn clone(&self) -> Self {
         if let Ruby::Data = self.ruby_type() {
-            unsafe {
-                let ptr = mem::transmute::<*mut c_void, Rc<RefCell<_>>>(self.value.value.p);
-                let cloned_ptr = Rc::clone(&ptr);
-                mem::forget(ptr);
-                Self::new(
-                    Rc::clone(&self.interp),
-                    sys::mrb_value {
-                        value: sys::mrb_value__bindgen_ty_1 {
-                            p: mem::transmute::<Rc<RefCell<_>>, *mut c_void>(cloned_ptr),
-                        },
-                        tt: self.value.tt
+            let ptr = unsafe { sys::mrb_sys_get_ptr(self.inner()) };
+            let cloned_ptr = ptr.clone();
```

Comment:
hmmm this is tricky. After thinking about this a bit, we need to transmute to `Rc` so we can increase the ref count of the `Rc` when cloning, but we don't know the `T` inside the `RefCell`. 🤔

Reply:
hmmm can we `match` on `T` if it must be one of a predefined list of possible types?

Reply:
@felipecsl I've done some noodling on this and I think we have a way forward here. We need to be generic on `T` but don't want to make `Value` generic as that would make the API too cumbersome.

Instead, we can:

- add a `Box<fn(Value) -> Value>` on the `Value` struct that implements clone.
- Add an associated function `Value::cloner::<T>` that can do the transmute in terms of a real type.
- Add a new Value constructor specifically for Rust-backed types that should be `Ruby::Data` that sets the cloner prop by `<T>`.

Reply:
hmm ok I think o got items 1 and 2, 3 is a bit unclear but we can keep iterating